### PR TITLE
Add WHISK fork only tests

### PR DESCRIPTION
**Pull request by dapplion**
### Description

Adds a baseline of tests covering the whisk feature introduced with https://github.com/ethereum/consensus-specs/pull/3342. In summary:
- generate and validate tracker, opening and shuffle proofs
- error cases for each assert condition in the spec

### Scope / future work

Tests are only run for the whisk fork and minimal. The current crypto backend is significantly slower than BLS and KZG, so running more tests than that is prohibitively expensive at the moment.

Also to run tests from previous forks on Whisk, the vector generation has to be modified to accommodate the fact that the proposer is not known, and requires validator_size factorial G1 multiplications in an implementation without revealed shuffling data to figure out who is the proposer for a whisk round.

All those points can be addressed, but in a future PR if necessary

### CODE REVIEW

File: tests/core/pyspec/eth2spec/test/context.py [hunk 1]
```diff
@@ -545,6 +549,7 @@ def wrapper(*args, spec: Spec, **kw):
 with_deneb_and_later = with_all_phases_from(DENEB)
 with_eip6110_and_later = with_all_phases_from(EIP6110)
 with_eip7002_and_later = with_all_phases_from(EIP7002)
+with_whisk_and_later = with_all_phases_from(WHISK)
```
Comment (hwwhww) [inline tests/core/pyspec/eth2spec/test/context.py:552]: Oops, so `with_all_phases_from` only iterates `ALL_PHASES`, which `WHISK` is not part of it now. Hence, Whisk state test cases are still skipped even with `--fork=whisk`. 😬
https://github.com/ethereum/consensus-specs/blob/5c194bb4934917b4bd65e6fb0ad3ce5e459e36af/tests/core/pyspec/eth2spec/test/context.py#L421-L427

A quick solution is adding a new argument `all_phases` to this helper:

```python
def with_all_phases_from(earliest_phase, all_phases=ALL_PHASES):
    """
    A decorator factory for running a tests with every phase except the ones listed
    """
    def decorator(fn):
        return with_phases([phase for phase in all_phases if is_post_fork(phase, earliest_phase)])(fn)
    return decorator
```

and re-write `with_whisk_and_later`:
```python
with_whisk_and_later = with_all_phases_from(WHISK, all_phases=ALLOWED_TEST_RUNNER_FORKS)
```

**Comment by dapplion** [comment]
Note:
- Bumped curdleproofs lib to use faster backend https://github.com/nalinbhardwaj/curdleproofs.pie/pull/19

**Comment by dapplion** [comment]
@hwwhww seems circle CI is experiencing a caching problem again https://app.circleci.com/pipelines/github/ethereum/consensus-specs/9434/workflows/86a7977e-ac75-42f6-8d2e-a81982c949e4/jobs/70233

```
ImportError while loading conftest '/home/circleci/specs-repo/tests/core/pyspec/eth2spec/test/conftest.py'.
eth2spec/test/conftest.py:1: in <module>
    from eth2spec.test import context
eth2spec/test/context.py:12: in <module>
    from eth2spec.whisk import mainnet as spec_whisk_mainnet, minimal as spec_whisk_minimal
eth2spec/whisk/__init__.py:1: in <module>
    from . import mainnet as spec  # noqa:F401
eth2spec/whisk/mainnet.py:222: in <module>
    CURDLEPROOFS_CRS = curdleproofs.CurdleproofsCrs.from_random_points(                 WHISK_VALIDATORS_PER_SHUFFLE,                 CURDLEPROOFS_N_BLINDERS,                 [G1Point.from_compressed_bytes_unchecked(p) for p in CURDLEPROOFS_CRS_G1]             )  # noqa: E501
E   AttributeError: type object 'CurdleproofsCrs' has no attribute 'from_random_points'
```

this method `from_random_points` exists in the referenced version and github CI and my local env pass tests

**Comment by hwwhww** [comment]
@dapplion 

Oh my, it's so weird. I pushed a commit in my branch: https://github.com/ethereum/consensus-specs/commit/ccd8026b10e03c813afb1b79792676c1ea10666a, and [it seems to have installed with the correct commit but has some other errors](https://app.circleci.com/pipelines/github/ethereum/consensus-specs/9446/workflows/db5a259a-f153-4f1f-8516-e5b549586ef5/jobs/70391).

But after I pushed that commit to this PR's branch (https://github.com/ethereum/consensus-specs/pull/3442/commits/ccd8026b10e03c813afb1b79792676c1ea10666a), it showed [the same wrong commit error here](https://app.circleci.com/pipelines/github/ethereum/consensus-specs/9446/workflows/db5a259a-f153-4f1f-8516-e5b549586ef5/jobs/70391).

Is it possible because your repository has a different CircleCI setting?

**Comment by dapplion** [comment]
> Is it possible because your repository has a different CircleCI setting?

The reported errors were correct, there was a type issue with curdleproof's return values not caught in my local env due to caching.

With https://github.com/nalinbhardwaj/curdleproofs.pie/pull/21 merged, now all tests should pass :)

**Comment by dapplion** [comment]
@hwwhww Published curdleproofs :tada: this PR is no longer using a Github ref

https://github.com/ethereum/consensus-specs/blob/0507b2730a854596c7c876e50114b279d4463783/setup.py#L537

File: specs/_features/whisk/beacon-chain.md [hunk 2]
```diff
@@ -83,14 +83,15 @@ def bytes_to_bls_field(b: Bytes32) -> BLSFieldElement:
...
+    return BLSFieldElement(field_element % BLS_MODULUS)
 ```
 
-| Name               | Value                                                                           |
-| ------------------ | ------------------------------------------------------------------------------- |
-| `BLS_G1_GENERATOR` | `bls.G1_to_bytes48(bls.G1)`                                                     |
-| `BLS_MODULUS`      | `52435875175126190479447740508185965837690552500527637822603658699938581184513` |
+| Name                  | Value                                                                           |
+| --------------------- | ------------------------------------------------------------------------------- |
+| `BLS_G1_GENERATOR`    | `Bytes48` |
```
Comment (mkalinin) [inline specs/_features/whisk/beacon-chain.md:91]: Is the value to be defined?

File: specs/_features/whisk/beacon-chain.md [hunk 3]
```diff
@@ -83,14 +83,15 @@ def bytes_to_bls_field(b: Bytes32) -> BLSFieldElement:
...
+| `BLS_MODULUS`         | `52435875175126190479447740508185965837690552500527637822603658699938581184513` |
+| `CURDLEPROOFS_CRS_G1` | `Vector[BLSG1Point, WHISK_VALIDATORS_PER_SHUFFLE + CURDLEPROOFS_N_BLINDERS + 3]`, contents TBD
```
Comment (mkalinin) [inline specs/_features/whisk/beacon-chain.md:93]: Consider naming a magic `3` number by introducing a constant

**Comment by dapplion** [comment]
Updated the CRS with correct serialization. Tests should pass now.

@hwwhww There's this error on CI which I can't reproduce locally. The error points to a spec variable not being present, however it's defined in the preset file. Any ideas?

```
 eth2spec/test/whisk/block_processing/test_process_shuffled_trackers.py:132: in test_invalid_gap_non_zero_proof
    set_state_epoch_selection_gap(spec, state)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

spec = <module 'eth2spec.whisk.minimal' from '/home/devops/consensus-specs-02/consensus-specs-02/consensus-specs/consensus-specs/tests/core/pyspec/eth2spec/whisk/minimal.py'>
state = BeaconState(Container)
    genesis_time: uint64 = 0
    genesis_validators_root: Root = 0x0a08c27fe4ece2483f9e581f78c6...e002bbe9af91c6de91ab693f6477551e7ee0a1e675d0fc614814d8a8aa
                                                           )

    def set_state_epoch_selection_gap(spec, state):
>       set_state_epoch(spec, state, spec.WHISK_EPOCHS_PER_SHUFFLING_PHASE - 1)
E       AttributeError: module 'eth2spec.whisk.minimal' has no attribute 'WHISK_EPOCHS_PER_SHUFFLING_PHASE'
```

**Review by mkalinin — APPROVED** [review]
The PR looks good to me!
@hwwhww what do you think about merging it after fixing the CI issue?

File: tests/core/pyspec/eth2spec/test/whisk/block_processing/test_process_shuffled_trackers.py [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+def get_pre_shuffle_trackers(spec, state, body):
+    return [state.whisk_candidate_trackers[i] for i in spec.get_shuffle_indices(body.randao_reveal)]
+
+
+def set_state_epoch(spec, state, epoch):
+    state.slot = epoch * spec.SLOTS_PER_EPOCH
+
+
+def set_state_epoch_selection_gap(spec, state):
+    set_state_epoch(spec, state, spec.WHISK_EPOCHS_PER_SHUFFLING_PHASE - 1)
```
Comment (hwwhww) [inline tests/core/pyspec/eth2spec/test/whisk/block_processing/test_process_shuffled_trackers.py:35]: `WHISK_EPOCHS_PER_SHUFFLING_PHASE` is in config, so here we should add `spec.config` in testing:
```suggestion
    set_state_epoch(spec, state, spec.config.WHISK_EPOCHS_PER_SHUFFLING_PHASE - 1)
```

File: tests/core/pyspec/eth2spec/test/whisk/block_processing/test_process_shuffled_trackers.py [hunk 5]
```diff
@@ -0,0 +1,130 @@
...
+    body = empty_block_body(spec)
+    set_state_epoch_selection_gap(spec, state)
+    yield from run_process_shuffled_trackers(spec, state, body)
+
+
+@with_whisk_and_later
+@spec_state_test
+def test_no_shuffle_minus_one_and_selection_gap(spec, state):
+    body = empty_block_body(spec)
+    set_state_epoch(spec, state, spec.WHISK_EPOCHS_PER_SHUFFLING_PHASE - spec.WHISK_PROPOSER_SELECTION_GAP - 1)
```
Comment (hwwhww) [inline tests/core/pyspec/eth2spec/test/whisk/block_processing/test_process_shuffled_trackers.py:76]: ditto
```suggestion
    set_state_epoch(
        spec,
        state,
        spec.config.WHISK_EPOCHS_PER_SHUFFLING_PHASE - spec.config.WHISK_PROPOSER_SELECTION_GAP - 1,
    )
```

**Review by hwwhww — COMMENTED** [review]
Hi @dapplion 

I believe it was because these parameters are defined in `Configurations`, so we use `spec.config.*` to access them.

File: tests/core/pyspec/eth2spec/test/helpers/keys.py [hunk 6]
```diff
@@ -5,13 +5,17 @@
...
-
 known_whisk_trackers = {}
 
 
 def register_known_whisk_tracker(k_r_G: bytes, index: int):
     known_whisk_trackers[k_r_G] = index
+
+
+def whisk_ks_initial(i: int):
+    return i
```
Comment (hwwhww) [inline tests/core/pyspec/eth2spec/test/helpers/keys.py:16]: perhaps add a TODO mark here

**Review by hwwhww — APPROVED** [review]
LGTM!

**Pull request merged**